src/meta/src/hummock/manager/checkpoint.rs

```diff
@@ -0,0 +1,125 @@
+use risingwave_pb::hummock::HummockVersionCheckpoint;
+
+use crate::hummock::error::Result;
+use crate::hummock::manager::{read_lock, write_lock};
+use crate::hummock::metrics_utils::trigger_stale_ssts_stat;
+use crate::hummock::HummockManager;
+use crate::storage::MetaStore;
+
+/// A hummock version checkpoint compacts previous hummock version delta logs, and stores stale
+/// objects from those delta logs.
+impl<S> HummockManager<S>
+where
+    S: MetaStore,
+{
+    pub(crate) async fn read_checkpoint(&self) -> Result<Option<HummockVersionCheckpoint>> {
+        // We `list` then `read`. Because from `read`'s error, we cannot tell whether it's "object
+        // not found" or other kind of error.
+        use prost::Message;
+        let metadata = self
+            .object_store
+            .list(&self.checkpoint_path)
+            .await?
+            .into_iter()
+            .filter(|o| o.key == self.checkpoint_path)
+            .collect_vec();
+        assert!(metadata.len() <= 1);
+        if metadata.is_empty() {
+            return Ok(None);
+        }
+        let data = self.object_store.read(&self.checkpoint_path, None).await?;
+        let ckpt = HummockVersionCheckpoint::decode(data).map_err(|e| anyhow::anyhow!(e))?;
+        Ok(Some(ckpt))
+    }
+
+    pub(super) async fn write_checkpoint(
+        &self,
+        checkpoint: &HummockVersionCheckpoint,
+    ) -> Result<()> {
+        use prost::Message;
+        let buf = checkpoint.encode_to_vec();
+        self.object_store
+            .upload(&self.checkpoint_path, buf.into())
+            .await?;
+        Ok(())
+    }
+
+    /// Creates a hummock version checkpoint.
+    /// Returns the diff between new and old checkpoint id.
+    /// Note that this method must not be called concurrently, because internally it doesn't hold
+    /// lock throughout the method.
+    #[named]
+    pub async fn create_version_checkpoint(&self, min_delta_log_num: u64) -> Result<u64> {
+        // 1. hold read lock and create new checkpoint
+        let versioning_guard = read_lock!(self, versioning).await;
+        let versioning = versioning_guard.deref();
+        let current_version = &versioning.current_version;
+        let old_checkpoint = &versioning.checkpoint;
+        let new_checkpoint_id = current_version.id;
+        let old_checkpoint_id = old_checkpoint.checkpoint.as_ref().unwrap().id;
+        if new_checkpoint_id < old_checkpoint_id + min_delta_log_num {
+            return Ok(0);
+        }
+        let mut checkpoint_version = old_checkpoint.checkpoint.as_ref().unwrap().clone();
+        let mut stale_objects = old_checkpoint.stale_objects.clone();
+        for (_, version_delta) in versioning
+            .hummock_version_deltas
+            .range((Excluded(old_checkpoint_id), Included(new_checkpoint_id)))
+        {
+            assert_eq!(version_delta.prev_id, checkpoint_version.id);
+            checkpoint_version.apply_version_delta(version_delta);
```

Comment:
can we use `current_version` as a new checkpoint instead of apply_version_delta  every time?

IIRC,  current_version is the lasted version of the checkpoint that what we need?

Reply:
Thanks. Fixed.